### PR TITLE
fix(2274): pass ssh agent to build container

### PR DESCRIPTION
**Pull request by wahapo**
## Context
fix https://github.com/screwdriver-cd/screwdriver/issues/2274

## Objective
Since macOS sockets do not seem to be available on the docker container, ~I decided to mount the ssh key~ instead we can use `/run/host-services/ssh-auth.sock` as the socket. On the linux machine connected with agent forwarding enabled, mounting `$SSH_AUTH_SOCK` solves the problem.

## References
https://github.com/screwdriver-cd/screwdriver/issues/2274
https://screwdriver-cd.slack.com/archives/C1T5T67PT/p1605035781056600
https://stackoverflow.com/questions/27036936/using-ssh-agent-with-docker-on-macos

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: launch/docker.go [hunk 1]
```diff
@@ -122,7 +122,7 @@ func (d *docker) runBuild(buildEntry buildEntry) error {
 	}
 
 	dockerCommandArgs := []string{"container", "run"}
-	dockerCommandOptions := []string{"--rm", "-v", srcVol, "-v", artVol, "-v", binVol, "-v", habVol, buildImage}
+	dockerCommandOptions := []string{"--rm", "-v", srcVol, "-v", artVol, "-v", binVol, "-v", habVol, "-v", fmt.Sprintf("%s/.ssh/:/root/.ssh/", os.Getenv("HOME")), "-v", fmt.Sprintf("%s:/tmp/auth.sock", os.Getenv("SSH_AUTH_SOCK")), "-e", "SSH_AUTH_SOCK=/tmp/auth.sock", buildImage}
```
Comment (jithine) [inline launch/docker.go:125]: can we call this `ssh-auth.sock` ?

**Review by jithine — COMMENTED** [review]
Docker for Mac has a weird self-provided ssh-agent https://github.com/docker/for-mac/issues/410#issuecomment-577064671

sort of works

```
docker run -it -v /run/host-services/ssh-auth.sock:/tmp/ssh-auth.sock -e SSH_AUTH_SOCK="/tmp/ssh-auth.sock"  node:12 ssh-add -l
The agent has no identities.
```

However this might not be the ssh agent user is running. 

I think I like this solution, only thing to figure out is whether in Mac, is it possible to get the ssh keys added to agent inside build container ?
Well for git clone use case, user can override the `GIT_SSH` variable.

**Comment by jithine** [comment]
Also I think it's beneficial to let user mount a custom folder. My use case is to use Athenz x509 certs, which are available in ~/.athenz folder. We can just allow users to mount custom folders to satisfy such user cases.

File: launch/docker.go (hunk 1 above)
Comment (wahapo) [reply launch/docker.go:125]: We can use `auth.sock` if there is no ssh keys. I think it's only effective when running sd-local on a remote server.

**Comment by wahapo** [comment]
We just need to be able to specify the directory to mount with the `-v` option (the same option as the docker).

**Comment by wahapo** [comment]
Is there a case for specifying `GIT_SSH` in screwdriver.yaml? I'm not sure, so I'd appreciate it if you could provide a concrete example.

**Comment by tanoda** [comment]
https://github.com/docker/for-mac/issues/410#issuecomment-577064671
The comment mentions about ssh-agent forwarding on Docker Desktop for Mac.

**Comment by jithine** [comment]
> Is there a case for specifying GIT_SSH in screwdriver.yaml? I'm not sure, so I'd appreciate it if you could provide a concrete example.

I was thinking about Mac. Since users ssh agent cannout be mounted inside, and we are mounting ssh keys folder, user can probably say

`sd-loca build job -e "GIT_SSH=ssh -i /root/my_rsa"`

**Comment by wahapo** [comment]
I think uses can also run it with something like `sd-local build job -e "GIT_SSH=ssh -i /root/.ssh/my_rsa"`. Is there any problem?

I tried it and it looks like we need to enter the password. This is likely to be a problem in the case of interactive mode. The concern is this?
```
$ cat ~/.ssh/run.sh
ssh -i /root/.ssh/id_rsa "$@"

$ sd-local build job -i --src-url git@<github_url>:<org>/<repo>.git -e "GIT_SSH=/root/.ssh/run.sh"

# git clone git@<github_url>:<org>/<repo>.git
Cloning into '<repo>'...
Enter passphrase for key '/root/.ssh/id_rsa':
```

**Comment by wahapo** [comment]
`.ssh` directory may be unnecessary. This might be fine.

```
jobs:
  job:
    image: node:12
    steps:
      - clone: |
          mkdir ~/.ssh/
          echo -e "Host github.com\n\tStrictHostKeyChecking no\n" >> ~/.ssh/config
          git clone git@github.com:<org>/<repo>.git
```

File: launch/docker.go [hunk 2]
```diff
@@ -122,7 +122,13 @@ func (d *docker) runBuild(buildEntry buildEntry) error {
 	}
 
 	dockerCommandArgs := []string{"container", "run"}
-	dockerCommandOptions := []string{"--rm", "-v", srcVol, "-v", artVol, "-v", binVol, "-v", habVol, buildImage}
+	sockFile := os.Getenv("SSH_AUTH_SOCK")
+	_, err = os.Stat("/run/host-services/ssh-auth.sock")
```
Comment (tanoda) [inline launch/docker.go:126]: FYI: This file exists in "Docker Desktop VM", not on macOS itself. So, I think there are no meanings to stat on this situation.

Comment (wahapo) [reply launch/docker.go:126]: Hmm, so this isn't working then.

**Comment by wahapo** [comment]
I guess this is the only way to go.

**Review by jithine — APPROVED** [review]
I think we should also add ability to mount custom volumes as well. 
I liked the idea of the mounting ssh keys, but I think we can leave that to users to use if they want, when in `interactive` mode

**Review by kumada626 — APPROVED** [review]

**Pull request merged**